### PR TITLE
#5118 - Affichage de la dernière diffusion mail et SMS lors d'une relance

### LINKED ISSUE #5118: ETQ prescripteur je souhaite connaître la date de dernière diffusion (mail et SMS) quand je fais la relance afin de ne pas spammer les utilisateurs
### :thinking: Contexte et problématique
En tant que prescripteur, il est nécessaire de connaître la date de la dernière relance par mail ou SMS réalisée directement dans la page de pilotage de convention. Ce besoin a été remonté principalement côté FT (France Travail).

Cela permettrait aussi d'éviter le spam auprès de l'entreprise (réprésentant légal ou tuteur), du bénéficiaire (et/ou du représentant légal). 

lien avec https://github.com/gip-inclusion/immersion-facile/issues/4767 et #4980.

> [!IMPORTANT]
> Attention à la recette :) 

#### 👉Actions concernées : 
* signature de la convention par le bénéficiaire
* signature de la convention par le représentant légal du bénéficiaire
* signature de la convention par le représentant légal de l'entreprise
* complétion du bilan par le tuteur
* signature du bilan par le bénéficiaire

#### 👷‍♂️Sur la page de pilotage d'une convention
**Etant donné** qu'une diffusion d'un mail (e premier mail envoyé, ou relance) a déjà eu lieu 
**Alors** je peux voir la date et l'heure de la dernière diffusion "Dernier mail ou SMS envoyé le JJ/MM/AAAA à HH:MM (heure française) à côté des CTA relances (encart bénéficiaire, entreprise, tuteur)

> [!WARNING]
> est-ce qu'on précise : a été relancé  "dernier mail" v.s "dernier SMS"

> [!TIP]
> aujourd'hui on a déjà la logique de comptabiliser à la fois le mail "originel" et les mails de relances

### :camera: Maquettes

Image
Image

[Fichier Figma](https://www.figma.com/design/QfpcTsc4KOf0P3xmXhb5n3/Composants-IF?node-id=271-3260&t=AlltufCxdyvKOYFz-0)

### 🤙 Data à suivre 

Voir l'impact

https://metabase-v2.immersion-facile.beta.gouv.fr/dashboard/17-tdb-suivi-delai-soumission-a-signature-hors-mini-stage

### PR DESCRIPTION
**Pull request by clement-duport**
Fixes #5118

### Revue du code (self-review)
- [x] Commits propres (pas de WIP, squash si nécessaire)
- [x] Relecture complète du diff effectuée
- [x] Pas de `console.log`, `TODO` ou code de debug restant

### État de la PR
- [x] Le titre respecte la convention : `#ID_ISSUE - Message clair et en français, compréhensible par quelqu'un du métier`
- [x] Auteurs assignés sur la PR et l'issue liée
- [x] Description du travail commentée sur l'issue (cf. `/document-issue`)
- [x] L'issue est en **RfR** dans le [projet GitHub](https://github.com/orgs/gip-inclusion/projects/10?query=sort%3Aupdated-desc+is%3Aopen)

### CODE REVIEW

File: shared/src/utils/date.unit.test.ts [hunk 1]
```diff
@@ -77,6 +77,20 @@ describe("Date utils tests - hours cooldown", () => {
...
+    ).toBe("24h00");
+  });
+
+  it("does not format minutes as 60", () => {
+    const lastActionAt = new Date(now.getTime() - 1000);
+
+    expect(
+      formatHoursCooldownTimeRemaining({ lastActionAt, minHours, now }),
+    ).toBe("24h00");
+  });
```
Comment (bbohec) [inline shared/src/utils/date.unit.test.ts:100]: Je comprends pas très bien ce test.
 const lastActionAt = new Date(now.getTime() - 1000); ça donne quelle date ?

File: shared/src/convention/convention.dto.ts [hunk 2]
```diff
@@ -499,9 +499,39 @@ export type ConventionAssessmentFields = {
     | null;
 };
 
+export type LastReminderDateByNotificationKind = {
+  email: DateTimeIsoString | null;
+  sms: DateTimeIsoString | null;
+};
+
```
Comment (bbohec) [inline shared/src/convention/convention.dto.ts:506]: Je met ça là mais c'est pas lié au code mais à la lecture de l'issue.
Pour le screen suivant que tu as mis, j'ai l'impression que le radio button est sélectionné bien que désactivé. Tu as vérifié que le submit n'est pas possible ? 

Image

File: shared/src/convention/convention.schema.ts [hunk 3]
```diff
@@ -580,6 +581,21 @@ export const conventionAssessmentFieldsSchema = z
   ])
   .nullable();
 
+const lastReminderDateByNotificationKindSchema = z.object({
+  email: dateTimeIsoStringSchema.nullable(),
+  sms: dateTimeIsoStringSchema.nullable(),
+});
```
Comment (bbohec) [inline shared/src/convention/convention.schema.ts]:
```suggestion
const lastReminderDateByNotificationKindSchema:ZodSchemaWithInputMatchingOutput<LastReminderDateByNotificationKind > = z.object({
  email: dateTimeIsoStringSchema.nullable(),
  sms: dateTimeIsoStringSchema.nullable(),
});
```

File: back/src/domains/convention/adapters/pgConventionSql.ts [hunk 4]
```diff
@@ -687,6 +696,137 @@ const getUsersWithAgencyRole = async (
...
+  },
+  assessmentSignature: {
+    email: "ASSESSMENT_NEEDS_SIGNATURE_BENEFICIARY_NOTIFICATION",
+    sms: "ReminderForAssessmentSignature",
+  },
+} as const;
+
+type ReminderKind = keyof typeof reminderKinds;
+
+export const getLastReminderDatesForRecipient = async ({
```
Comment (bbohec) [inline back/src/domains/convention/adapters/pgConventionSql.ts]: sur toutes les fonctions ici et en dessous, retirer les export.
Normalement on est dédié au PgConvention ici sauf peut être `getLastRemindersFieldsByConventions`

File: back/src/domains/convention/adapters/InMemoryConventionQueries.ts [hunk 5]
```diff
@@ -367,6 +368,7 @@ export class InMemoryConventionQueries implements ConventionQueries {
           }
         : undefined,
       ...assesmentEntityToConventionAssessmentFields(assessment),
+      lastReminders: makeEmptyLastReminders(),
```
Comment (bbohec) [inline back/src/domains/convention/adapters/InMemoryConventionQueries.ts:371]: C'était ce que tu m'avais dit ?
On le fait pour le PG uniquement pour l'instant mais c'est temporaire car on va retravailler les repos ?

File: back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts [hunk 6]
```diff
@@ -64,6 +66,205 @@ describe("Pg implementation of ConventionQueries", () => {
...
     .buildUser();
 
+  const insertEmailReminder = async ({
+    conventionId,
+    emailKind,
+    recipientEmail,
+    createdAt,
+  }: {
+    conventionId: ConventionId;
+    emailKind: string;
```
Comment (bbohec) [inline back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts]: utiliser EmailKind ou assimilé

File: back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts [hunk 7]
```diff
@@ -64,6 +66,205 @@ describe("Pg implementation of ConventionQueries", () => {
...
+  };
+
+  const insertSmsReminder = async ({
+    conventionId,
+    smsKind,
+    recipientPhone,
+    createdAt,
+  }: {
+    conventionId: ConventionId;
+    smsKind: string;
```
Comment (bbohec) [inline back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts]: Idem reprendre le SMS kind au lieu de string

File: back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts [hunk 8]
```diff
@@ -64,6 +66,205 @@ describe("Pg implementation of ConventionQueries", () => {
     .withId(uuid())
     .buildUser();
 
+  const insertEmailReminder = async ({
```
Comment (bbohec) [inline back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts]: Peut être déplacer tout en bas du fichier ces fonctions utilitaires car ça pollue pas mal.

File: back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts [hunk 9]
```diff
@@ -196,6 +398,212 @@ describe("Pg implementation of ConventionQueries", () => {
...
+          recipientEmail: "benef@r.com",
+          createdAt: newerDate,
+        });
+
+        const result = await conventionQueries.getConventionById(conventionIdA);
+
+        expectToEqual(result?.lastReminders.conventionSignatures.beneficiary, {
+          email: newerDate.toISOString(),
+          sms: null,
+        });
```
Comment (bbohec) [inline back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts]: Je pense que c'est plus pertinant de tester l'entièreté de `lastReminders `afin de s'assurer qu'on a pas de side effect comme remplir d'autres props non souhaitées

File: back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts [hunk 10]
```diff
@@ -196,6 +398,212 @@ describe("Pg implementation of ConventionQueries", () => {
...
+
+      it("returns signature reminders for each signatory role", async () => {
+        const establishmentRepDate = new Date("2025-01-02T12:00:00.000Z");
+        const beneficiaryRepDate = new Date("2025-01-02T13:00:00.000Z");
+        const currentEmployerDate = new Date("2025-01-02T14:00:00.000Z");
+
+        await insertEmailReminder({
+          conventionId: conventionIdA,
+          emailKind: "NEW_CONVENTION_CONFIRMATION_REQUEST_SIGNATURE",
+          recipientEmail: "est@rep.com",
```
Comment (bbohec) [inline back/src/domains/convention/adapters/PgConventionQueries.integration.test.ts]: reprendre les emails à partir de la convention pour éviter les valeurs en dur

File: front/src/app/components/convention/SendReminderModal.tsx [hunk 11]
```diff
@@ -15,6 +21,43 @@ type ReminderModal = {
...
+}): DateTimeIsoString => {
+  if (smsLastSentAt) {
+    if (new Date(emailLastSentAt) <= new Date(smsLastSentAt))
+      return emailLastSentAt;
+
+    return smsLastSentAt;
+  }
+
+  return emailLastSentAt;
+};
```
Comment (bbohec) [inline front/src/app/components/convention/SendReminderModal.tsx]: un peu trop d'imbrication de if.


```suggestion
const getFirstChannelAvailableAgainLastSentAt = ({
  emailLastSentAt,
  smsLastSentAt,
}: {
  emailLastSentAt: DateTimeIsoString;
  smsLastSentAt?: DateTimeIsoString | null;
}): DateTimeIsoString => {
  if (!smsLastSentAt) return emailLastSentAt;
  return new Date(emailLastSentAt) <= new Date(smsLastSentAt)
    ? emailLastSentAt
    : smsLastSentAt
};
```

File: front/src/app/components/convention/SendReminderModal.tsx [hunk 12]
```diff
@@ -15,6 +21,43 @@ type ReminderModal = {
...
+    date: lastActionAt,
+    withHours: true,
+  })}. Afin d’éviter les envois excessifs, vous ne pourrez effectuer un nouvel envoi que dans ${formatHoursCooldownTimeRemaining(
+    {
+      lastActionAt,
+      minHours: CONVENTION_MANUAL_REMINDER_COOLDOWN_IN_HOURS,
+      now,
+    },
+  )}.`;
+};
```
Comment (bbohec) [inline front/src/app/components/convention/SendReminderModal.tsx]: Idem déplacer les fonctions utilitaires en bas du fichier pour avoir accès de base au composant.

C'est plus simple pour la relecture du composant d'avoir:
```
COMPOSANT
util()
util2()
```
que
```
util()
util2()
COMPOSANT
```

**Review by bbohec — CHANGES_REQUESTED** [review]

File: back/src/domains/convention/adapters/InMemoryConventionQueries.ts (hunk 5 above)
Comment (clement-duport) [reply back/src/domains/convention/adapters/InMemoryConventionQueries.ts:371]: yep

File: shared/src/utils/date.unit.test.ts (hunk 1 above)
Comment (clement-duport) [reply shared/src/utils/date.unit.test.ts:100]: C'est pour sécuriser un edge case d’arrondi : à 23h59m59s restants, on doit afficher 24h00 et pas 23h60.

J’ai clarifié le test en le remplaçant par subSeconds(now, 1), c’est plus explicite.

File: shared/src/convention/convention.dto.ts (hunk 2 above)
Comment (clement-duport) [reply shared/src/convention/convention.dto.ts:506]: non le screen shot n'est pas a jour. Je m'en était rendus compte entre temps.

File: shared/src/utils/date.unit.test.ts (hunk 1 above)
Comment (bbohec) [reply shared/src/utils/date.unit.test.ts:100]: nickel

**Review by bbohec — APPROVED** [review]
